pkg/stream/http/stream.go

```diff
@@ -302,26 +311,350 @@ func (conn *clientStreamConnection) serve() {
 			return
 		}
 
-		if log.Proxy.GetLogLevel() >= log.DEBUG {
-			log.Proxy.Debugf(s.stream.ctx, "[stream] [http] receive response, requestId = %v", s.stream.id)
+		// 1. handle response
+		if conn.useStream || httpRspUseStream {
+			handleStreamResponse(conn)
+		} else {
+			handleBlockedResponse(conn)
 		}
-
-		// 2. response processing
+		// 2. set client stream reset flag
 		resetConn := false
 		if s.response.ConnectionClose() {
 			resetConn = true
 		}
-
 		// 3. local reset if header 'Connection: close' exists
 		if resetConn {
 			// goaway the connpool
 			s.connection.streamConnectionEventListener.OnGoAway()
 		}
+	}
+}
 
-		if atomic.LoadInt32(&s.readDisableCount) <= 0 {
-			s.handleResponse()
+// handleStreamResponse: http stream response
+func handleStreamResponse(conn *clientStreamConnection) {
+	s := conn.stream
+	startStreamResponse := func(cs *clientStream) {
+		_ = variable.Set(s.ctx, types.VarHttpResponseUseStream, true)
+		s.connection.mutex.Lock()
+		s.connection.stream = nil
+		s.connection.mutex.Unlock()
+		cs.recData = buffer.NewPipeBuffer(0)
+		cs.receiver.OnReceive(cs.ctx, mosnhttp.ResponseHeader{ResponseHeader: &s.response.Header}, cs.recData, nil)
+	}
+
+	sendStreamResponse := func(cs *clientStream) error {
+		return writeBodyToPipe(conn.br, 0, s.response.Header.ContentLength(), func(data []byte) error {
+			if _, err := cs.recData.Write(data); err != nil {
+				return err
+			}
+			return nil
+		})
+	}
+	finishStreamResponse := func(cs *clientStream) {
+		cs.recData.CloseWithError(io.EOF)
+		// destroy stream
+		cs.stream.DestroyStream()
+	}
+
+	startStreamResponse(s)
+	if err := sendStreamResponse(s); err != nil {
+		log.Proxy.Errorf(s.connection.context, "[stream] [http] [stream response] client stream write buffer: %s", err)
+		reason := conn.resetReason
+		if reason == "" {
+			reason = types.StreamRemoteReset
 		}
+		s.ResetStream(reason)
+		return
 	}
+	finishStreamResponse(s)
+}
+
+// handleBlockedResponse: http blocked response
+func handleBlockedResponse(conn *clientStreamConnection) {
+	s := conn.stream
+	_ = variable.Set(s.ctx, types.VarHttpResponseUseStream, false)
+
+	err := s.response.ReadBody(conn.br, 0)
+	if err != nil {
+		if s != nil {
+			log.Proxy.Errorf(s.connection.context, "[stream] [http] client stream connection wait response error: %s", err)
+			reason := conn.resetReason
+			if reason == "" {
+				reason = types.StreamRemoteReset
+			}
+			s.ResetStream(reason)
+		}
+		return
+	}
+
+	if log.Proxy.GetLogLevel() >= log.DEBUG {
+		log.Proxy.Debugf(s.stream.ctx, "[stream] [http] receive response, requestId = %v", s.stream.id)
+	}
+
+	if atomic.LoadInt32(&s.readDisableCount) <= 0 {
+		s.handleResponse()
+	}
+}
+
+// copy from fasthttp github.com/valyala/fasthttp@v1.40.0/http.go:1374
+func writeBodyToPipe(r *bufio.Reader, maxBodySize int, contentLength int, writeFunc func([]byte) error) (err error) {
+	buf := make([]byte, 0)
```

Comment:
Perf: give a initial capacity to avoid frequent growth

Reply:
ok